{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66d21ef5-55e3-4e63-b1c4-6c23df47434d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-27T01:29:01.108849Z",
     "iopub.status.busy": "2025-11-27T01:29:01.108577Z",
     "iopub.status.idle": "2025-11-27T01:29:36.444507Z",
     "shell.execute_reply": "2025-11-27T01:29:36.444036Z",
     "shell.execute_reply.started": "2025-11-27T01:29:01.108830Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking tasks to skip...\n",
      "Checking pending tasks to submit...\n",
      "Total pending pids available to submit: 0 / 0\n"
     ]
    }
   ],
   "source": [
    "# rmstools setup_wizard -dbpath /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/test_DB/ -dbname test.db\n",
    "\n",
    "DBPATH = \"/fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/test_DB/\"\n",
    "DBNAME = \"test.db\"\n",
    "DBRESDIR = \"RMSResources/\"\n",
    "import sys\n",
    "import logging\n",
    "logger = logging.getLogger(\"rmspool\")\n",
    "logger.setLevel(\"INFO\")\n",
    "logger.addHandler(logging.StreamHandler(sys.stdout))\n",
    "from rmsp import rmscore\n",
    "rms = rmscore.ResourceManagementSystem(f\"{DBPATH}{DBNAME}\", f\"{DBPATH}{DBRESDIR}\")\n",
    "from rmsp import rmsbuilder\n",
    "rmspool = rmsbuilder.RMSProcessWrapPool(rms, 8)\n",
    "rmsb = rmsbuilder.RMSUnrunTasksBuilder(rmspool)\n",
    "from rmsp import rmsutils\n",
    "# This ID is randomly generated to trace Jupyter files in RMS\n",
    "rms.set_scriptID(\"bcc3719729074f69a93a31a4b58e9fda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "daca1269-c7da-4e25-a145-a52086ae73e0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-27T01:29:36.445617Z",
     "iopub.status.busy": "2025-11-27T01:29:36.445299Z",
     "iopub.status.idle": "2025-11-27T01:30:08.945204Z",
     "shell.execute_reply": "2025-11-27T01:30:08.943824Z",
     "shell.execute_reply.started": "2025-11-27T01:29:36.445598Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from commonhelper import load_func\n",
    "from commandtemplate.conda import run_template_bash, OUTPUT_run_template_bash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1aa344d1-e15f-420c-91ae-0f2c960364d0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T18:21:55.727931Z",
     "iopub.status.busy": "2025-11-04T18:21:55.727573Z",
     "iopub.status.idle": "2025-11-04T18:22:03.181132Z",
     "shell.execute_reply": "2025-11-04T18:22:03.180617Z",
     "shell.execute_reply.started": "2025-11-04T18:21:55.727909Z"
    }
   },
   "outputs": [],
   "source": [
    "conda_env = \"DI\"\n",
    "pipeline = \"./procapanalysis.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "080dbf0c-22a9-465d-bfaa-335f56ddf98c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-27T01:30:08.946607Z",
     "iopub.status.busy": "2025-11-27T01:30:08.946061Z",
     "iopub.status.idle": "2025-11-27T01:30:08.949346Z",
     "shell.execute_reply": "2025-11-27T01:30:08.948929Z",
     "shell.execute_reply.started": "2025-11-27T01:30:08.946588Z"
    }
   },
   "outputs": [],
   "source": [
    "PROJECT_DIR_d = \"/fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/PROcap/\"\n",
    "PROJECT_DIR_r = \"/fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/\"\n",
    "PROJECT_DIR_o = \"/fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/output/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e903f004-a3f7-452a-b1ad-bd382a86b4c3",
   "metadata": {},
   "source": [
    "# Pipeline registration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e865730b-757f-41a8-a051-c3bad367d7a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T18:22:05.891743Z",
     "iopub.status.busy": "2025-11-04T18:22:05.891412Z",
     "iopub.status.idle": "2025-11-04T18:22:08.752807Z",
     "shell.execute_reply": "2025-11-04T18:22:08.752274Z",
     "shell.execute_reply.started": "2025-11-04T18:22:05.891723Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "run_template_bash = rms.register_pipe(run_template_bash, output_func = OUTPUT_run_template_bash)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7004f45a-5bc3-41ad-843f-0eeb334b383c",
   "metadata": {},
   "source": [
    "# Raw PRO-cap data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "57d25ef2-da76-44ba-90f5-9c2463030645",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-27T01:31:05.357570Z",
     "iopub.status.busy": "2025-11-27T01:31:05.357156Z",
     "iopub.status.idle": "2025-11-27T01:31:05.375123Z",
     "shell.execute_reply": "2025-11-27T01:31:05.374685Z",
     "shell.execute_reply.started": "2025-11-27T01:31:05.357549Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library Name</th>\n",
       "      <th>Internal Sample ID</th>\n",
       "      <th>Cell line</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>S2 spike-in</th>\n",
       "      <th>5' adapter</th>\n",
       "      <th>3' adapter</th>\n",
       "      <th>Sequencing Read 1</th>\n",
       "      <th>Sequencing Read 2</th>\n",
       "      <th>ENCODE Experiment ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C1a</td>\n",
       "      <td>C1a</td>\n",
       "      <td>K562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>C1a_CKDL220028609-1A_HLJNNDSX5_L4_1.fq.gz</td>\n",
       "      <td>C1a_CKDL220028609-1A_HLJNNDSX5_L4_2.fq.gz</td>\n",
       "      <td>ENCSR220XSM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C1b</td>\n",
       "      <td>C1b</td>\n",
       "      <td>K562</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>C1b_CKDL220028609-1A_HLJNNDSX5_L4_1.fq.gz</td>\n",
       "      <td>C1b_CKDL220028609-1A_HLJNNDSX5_L4_2.fq.gz</td>\n",
       "      <td>ENCSR220XSM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCF_U1</td>\n",
       "      <td>CTCF_U1</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>untreated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>CTCF_U1_1.fq.gz</td>\n",
       "      <td>CTCF_U1_2.fq.gz</td>\n",
       "      <td>ENCSR165QKR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCF_U2</td>\n",
       "      <td>CTCF_U2</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>untreated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>CTCF_U2_1.fq.gz</td>\n",
       "      <td>CTCF_U2_2.fq.gz</td>\n",
       "      <td>ENCSR165QKR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTCF_T1</td>\n",
       "      <td>CTCF_T1</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>treated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>CTCF_T1_1.fq.gz</td>\n",
       "      <td>CTCF_T1_2.fq.gz</td>\n",
       "      <td>ENCSR261OWE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CTCF_T2</td>\n",
       "      <td>CTCF_T2</td>\n",
       "      <td>HCT116</td>\n",
       "      <td>treated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>CTCF_T2_1.fq.gz</td>\n",
       "      <td>CTCF_T2_2.fq.gz</td>\n",
       "      <td>ENCSR261OWE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ZNF143_U1</td>\n",
       "      <td>ZNF1</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>untreated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>ZNF1_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz</td>\n",
       "      <td>ZNF1_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz</td>\n",
       "      <td>ENCSR473KVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZNF143_U2</td>\n",
       "      <td>ZNF2</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>untreated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>ZNF2_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz</td>\n",
       "      <td>ZNF2_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz</td>\n",
       "      <td>ENCSR473KVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZNF143_T1</td>\n",
       "      <td>ZNF5</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>treated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>ZNF5_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz</td>\n",
       "      <td>ZNF5_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz</td>\n",
       "      <td>ENCSR733DAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ZNF143_T2</td>\n",
       "      <td>ZNF6</td>\n",
       "      <td>HEK293T</td>\n",
       "      <td>treated</td>\n",
       "      <td>Yes</td>\n",
       "      <td>VRA5 + 6N</td>\n",
       "      <td>VRA3 + 6N</td>\n",
       "      <td>ZNF6_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz</td>\n",
       "      <td>ZNF6_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz</td>\n",
       "      <td>ENCSR733DAS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Library Name Internal Sample ID Cell line  Treatment S2 spike-in 5' adapter  \\\n",
       "0          C1a                C1a      K562        NaN          No  VRA5 + 6N   \n",
       "1          C1b                C1b      K562        NaN          No  VRA5 + 6N   \n",
       "2      CTCF_U1            CTCF_U1    HCT116  untreated         Yes  VRA5 + 6N   \n",
       "3      CTCF_U2            CTCF_U2    HCT116  untreated         Yes  VRA5 + 6N   \n",
       "4      CTCF_T1            CTCF_T1    HCT116    treated         Yes  VRA5 + 6N   \n",
       "5      CTCF_T2            CTCF_T2    HCT116    treated         Yes  VRA5 + 6N   \n",
       "6    ZNF143_U1               ZNF1   HEK293T  untreated         Yes  VRA5 + 6N   \n",
       "7    ZNF143_U2               ZNF2   HEK293T  untreated         Yes  VRA5 + 6N   \n",
       "8    ZNF143_T1               ZNF5   HEK293T    treated         Yes  VRA5 + 6N   \n",
       "9    ZNF143_T2               ZNF6   HEK293T    treated         Yes  VRA5 + 6N   \n",
       "\n",
       "  3' adapter                           Sequencing Read 1  \\\n",
       "0  VRA3 + 6N   C1a_CKDL220028609-1A_HLJNNDSX5_L4_1.fq.gz   \n",
       "1  VRA3 + 6N   C1b_CKDL220028609-1A_HLJNNDSX5_L4_1.fq.gz   \n",
       "2  VRA3 + 6N                             CTCF_U1_1.fq.gz   \n",
       "3  VRA3 + 6N                             CTCF_U2_1.fq.gz   \n",
       "4  VRA3 + 6N                             CTCF_T1_1.fq.gz   \n",
       "5  VRA3 + 6N                             CTCF_T2_1.fq.gz   \n",
       "6  VRA3 + 6N  ZNF1_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz   \n",
       "7  VRA3 + 6N  ZNF2_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz   \n",
       "8  VRA3 + 6N  ZNF5_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz   \n",
       "9  VRA3 + 6N  ZNF6_CKDL250028480-1A_22Y7T3LT4_L6_1.fq.gz   \n",
       "\n",
       "                            Sequencing Read 2 ENCODE Experiment ID  \n",
       "0   C1a_CKDL220028609-1A_HLJNNDSX5_L4_2.fq.gz          ENCSR220XSM  \n",
       "1   C1b_CKDL220028609-1A_HLJNNDSX5_L4_2.fq.gz          ENCSR220XSM  \n",
       "2                             CTCF_U1_2.fq.gz          ENCSR165QKR  \n",
       "3                             CTCF_U2_2.fq.gz          ENCSR165QKR  \n",
       "4                             CTCF_T1_2.fq.gz          ENCSR261OWE  \n",
       "5                             CTCF_T2_2.fq.gz          ENCSR261OWE  \n",
       "6  ZNF1_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz          ENCSR473KVF  \n",
       "7  ZNF2_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz          ENCSR473KVF  \n",
       "8  ZNF5_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz          ENCSR733DAS  \n",
       "9  ZNF6_CKDL250028480-1A_22Y7T3LT4_L6_2.fq.gz          ENCSR733DAS  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We've deposited PRO-cap data in the ENCODE portal\n",
    "\n",
    "procap_info = pd.read_excel(f\"{PROJECT_DIR_o}tables/SuppTable1.xlsx\")\n",
    "procap_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b69d6b98-92f5-4b4b-9002-068f1795d7bd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T17:42:48.011016Z",
     "iopub.status.busy": "2025-11-04T17:42:48.010663Z",
     "iopub.status.idle": "2025-11-04T17:43:33.341433Z",
     "shell.execute_reply": "2025-11-04T17:43:33.340790Z",
     "shell.execute_reply.started": "2025-11-04T17:42:48.010994Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "files = []\n",
    "for _, row in procap_info.iterrows():\n",
    "\tif type(row[\"Sequencing Read 1\"]) == str:\n",
    "\t\tfiles.extend([row[\"Sequencing Read 1\"], row[\"Sequencing Read 2\"]])\n",
    "for f in files:\n",
    "\trms.register_file(f\"{PROJECT_DIR_d}/RawData/{f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05335602-e1da-4438-b6fa-846ba9922290",
   "metadata": {},
   "source": [
    "# Genome and related files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "51517856-9873-4f0e-bd93-ed4340e2955a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T17:14:26.073849Z",
     "iopub.status.busy": "2025-11-04T17:14:26.073498Z",
     "iopub.status.idle": "2025-11-04T17:14:26.138526Z",
     "shell.execute_reply": "2025-11-04T17:14:26.138103Z",
     "shell.execute_reply.started": "2025-11-04T17:14:26.073826Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "for f in [\n",
    "\t# https://www.encodeproject.org/files/GRCh38_no_alt_analysis_set_GCA_000001405.15/@@download/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz\", \n",
    "\t# NCBI GenBank ID U13369.1; https://www.ncbi.nlm.nih.gov/nuccore/555853\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/rDNA_U13369.1.fasta\",\n",
    "\t# https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/hg38.chrom.sizes.filtered\",\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/hg38.chrom.sizes.filtered.no.chrY\",\n",
    "\t# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/gencode.v37.annotation.gtf.gz\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\",\n",
    "\t# https://hgdownload.soe.ucsc.edu/goldenPath/dm6/bigZips/dm6.fa.gz\n",
    "\tf\"{PROJECT_DIR_r}genomes/drosophila/dm6.fa.gz\",\n",
    "\t# https://github.com/Boyle-Lab/Blacklist/raw/master/lists/hg38-blacklist.v2.bed.gz\n",
    "\tf\"{PROJECT_DIR_r}genomes/human/hg38-blacklist.v2.bed.gz\"\n",
    "]:\n",
    "\trms.register_file(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e5cc40a-a84b-4d5c-b533-8c916a8c022e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T17:14:26.594984Z",
     "iopub.status.busy": "2025-11-04T17:14:26.594748Z",
     "iopub.status.idle": "2025-11-04T17:14:33.980619Z",
     "shell.execute_reply": "2025-11-04T17:14:33.980196Z",
     "shell.execute_reply.started": "2025-11-04T17:14:26.594965Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fdfb7e63c40>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Unzip files\n",
    "\n",
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"zcat {i} > {o}\",\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta\"\n",
    ")\n",
    "run_template_bash(\n",
    "\t\"zcat {i} > {o}\",\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf\"\n",
    ")\n",
    "run_template_bash(\n",
    "\t\"zcat {i} > {o}\",\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/drosophila/dm6.fa.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/drosophila/dm6.fa\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5367163d-5b37-472a-a972-bce6636a365f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-11-04T17:14:57.909177Z",
     "iopub.status.busy": "2025-11-04T17:14:57.908855Z",
     "iopub.status.idle": "2025-11-04T17:15:13.931210Z",
     "shell.execute_reply": "2025-11-04T17:15:13.930685Z",
     "shell.execute_reply.started": "2025-11-04T17:14:57.909154Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Add prefix to dm6 for spike-in indication\n",
    "\n",
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"biodatatools modify_fasta_names -i {i} -o {o} -func {func}\",\n",
    "\tconda_env = conda_env,\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/drosophila/dm6.fa\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/drosophila/dm6_addprefix.fa\",\n",
    "\tfunc = \"'lambda x: \\\"dm6_\\\" + x'\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09d53913-165d-4a86-a590-18ca11d33c41",
   "metadata": {},
   "source": [
    "# STAR genome index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5434800-53ad-4c01-92bb-bcf1dd2d84b7",
   "metadata": {},
   "source": [
    "## Human without spikein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "944c5caf-485c-4a58-9e08-b3d1072460a1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T21:02:50.959855Z",
     "iopub.status.busy": "2024-10-11T21:02:50.958961Z",
     "iopub.status.idle": "2024-10-11T21:42:39.298990Z",
     "shell.execute_reply": "2024-10-11T21:42:39.297690Z",
     "shell.execute_reply.started": "2024-10-11T21:02:50.959806Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "!!!!! WARNING: Could not move Log.out file from ./Log.out into /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/STAR_genome_index/hg38_rDNA/Log.out. Will keep ./Log.out\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/home/yc2553/miniconda3_new/envs/DI_test/bin/STAR-avx2 --runMode genomeGenerate --readFilesCommand zcat --runThreadN 32 --genomeDir /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/STAR_genome_index/hg38_rDNA --genomeFastaFiles /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/genomes/human/rDNA_U13369.1.fasta --sjdbGTFfile /fs/cbsuhy02/storage/yc2553/yc2553/projects/TRE_directionality/resources/genomes/human/gencode.v37.annotation.gtf\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 11 17:02:53 ..... started STAR run\n",
      "Oct 11 17:02:53 ... starting to generate Genome files\n",
      "Oct 11 17:04:28 ..... processing annotations GTF\n",
      "Oct 11 17:05:28 ... starting to sort Suffix Array. This may take a long time...\n",
      "Oct 11 17:06:02 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Oct 11 17:26:29 ... loading chunks from disk, packing SA...\n",
      "Oct 11 17:28:22 ... finished generating suffix array\n",
      "Oct 11 17:28:22 ... generating Suffix Array index\n",
      "Oct 11 17:34:07 ... completed Suffix Array index\n",
      "Oct 11 17:34:08 ..... inserting junctions into the genome indices\n",
      "Oct 11 17:41:41 ... writing Genome to disk ...\n",
      "Oct 11 17:41:44 ... writing Suffix Array to disk ...\n",
      "Oct 11 17:42:31 ... writing SAindex to disk\n",
      "Oct 11 17:42:38 ..... finished successfully\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fea46303b50>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# C1 library: no spike-in\n",
    "\n",
    "f = lambda **kwargs: [kwargs['genomeDir']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"STAR --runMode genomeGenerate --readFilesCommand zcat --runThreadN {thread} --genomeDir {genomeDir} --genomeFastaFiles {i0} {i1} --sjdbGTFfile {annotation_file}\",\n",
    "\tconda_env = conda_env,\n",
    "\toutput_func = output_func,\n",
    "\ti0 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta\"),\n",
    "\ti1 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/rDNA_U13369.1.fasta\"),\n",
    "\tannotation_file = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf\"),\n",
    "\tthread = 32,\n",
    "\tgenomeDir = f\"{PROJECT_DIR_r}STAR_genome_index/hg38_rDNA\"\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f1a824e-4242-440c-badc-478fd1710aa7",
   "metadata": {},
   "source": [
    "## Human with drosophila spikein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52653479-d51b-4c77-bd03-6a519ad47152",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2025-11-03T15:07:19.232603Z",
     "iopub.status.idle": "2025-11-03T15:07:19.232916Z",
     "shell.execute_reply": "2025-11-03T15:07:19.232811Z",
     "shell.execute_reply.started": "2025-11-03T15:07:19.232801Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Degron lines: with spike-in\n",
    "\n",
    "f = lambda **kwargs: [kwargs['genomeDir']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"STAR --runMode genomeGenerate --readFilesCommand zcat --runThreadN {thread} --genomeDir {genomeDir} --genomeFastaFiles {i0} {i1} {i2} --sjdbGTFfile {annotation_file}\",\n",
    "\tconda_env = conda_env,\n",
    "\toutput_func = output_func,\n",
    "\ti0 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta\"),\n",
    "\ti1 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/rDNA_U13369.1.fasta\"),\n",
    "\ti2 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/drosophila/dm6_addprefix.fa\"),\n",
    "\tannotation_file = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf\"),\n",
    "\tthread = 32,\n",
    "\tgenomeDir = f\"{PROJECT_DIR_r}STAR_genome_index/hg38_rDNA_dm6\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e696e996-0e78-4886-ab54-68a922633506",
   "metadata": {},
   "source": [
    "# GENCODE annotation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bfb2cdb2-9935-48ec-9b40-4181de06e973",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T22:27:28.280538Z",
     "iopub.status.busy": "2024-10-11T22:27:28.278967Z",
     "iopub.status.idle": "2024-10-11T22:31:35.449728Z",
     "shell.execute_reply": "2024-10-11T22:31:35.448631Z",
     "shell.execute_reply.started": "2024-10-11T22:27:28.280471Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fea462ff7f0>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This file is used to classify distal vs. proximal elements\n",
    "\n",
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"biodatatools generate_union_TSS -i {i} -o {o} -forward_len {forward_len} -reverse_len {reverse_len}\",\n",
    "\tconda_env = conda_env,\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.TSS.500.bed.bgz\",\n",
    "\tforward_len = 500,\n",
    "\treverse_len = 500\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "74f0545b-03af-493d-a135-da165a75206f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T22:31:35.451833Z",
     "iopub.status.busy": "2024-10-11T22:31:35.451125Z",
     "iopub.status.idle": "2024-10-11T22:35:31.530551Z",
     "shell.execute_reply": "2024-10-11T22:35:31.529431Z",
     "shell.execute_reply.started": "2024-10-11T22:31:35.451782Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fea4657b5e0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This file is used to classify intergenic vs. intragenic elements\n",
    "\n",
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"biodatatools generate_union_transcripts -i {i} -o {o}\",\n",
    "\tconda_env = conda_env,\n",
    "\toutput_func = output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.transcripts.union.bed.bgz\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "18f695eb-a28a-4fd9-86e2-ffde8dcc4af4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T22:35:31.536840Z",
     "iopub.status.busy": "2024-10-11T22:35:31.534437Z",
     "iopub.status.idle": "2024-10-11T22:39:45.261242Z",
     "shell.execute_reply": "2024-10-11T22:39:45.260182Z",
     "shell.execute_reply.started": "2024-10-11T22:35:31.536744Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fea463012d0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"biodatatools generate_geneannotations_union_features -i {i} -o {o} -feature {f}\",\n",
    "\tconda_env=conda_env,\n",
    "\toutput_func=output_func,\n",
    "\ti = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.CDS.union.bed.bgz\",\n",
    "\tf = \"CDS\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a96aebd9-a22c-4bf2-be91-af7b4e087739",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-17T01:42:27.433909Z",
     "iopub.status.busy": "2024-10-17T01:42:27.433559Z",
     "iopub.status.idle": "2024-10-17T01:42:52.852362Z",
     "shell.execute_reply": "2024-10-17T01:42:52.851907Z",
     "shell.execute_reply.started": "2024-10-17T01:42:27.433884Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<rmsp.rmscore.Resource at 0x7fea4629ea40>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Exclude coding or blacklist regions\n",
    "# Used to generate control regions later\n",
    "\n",
    "f = lambda **kwargs: [kwargs['o']]\n",
    "output_func = rms.register_pipe(f)\n",
    "run_template_bash(\n",
    "\t\"zcat {i0} {i1} | cut -f 1-3 | gzip > {o}\",\n",
    "\tconda_env=conda_env,\n",
    "\toutput_func=output_func,\n",
    "\ti0 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.CDS.union.bed.bgz\"),\n",
    "\ti1 = rms.file_from_path(f\"{PROJECT_DIR_r}genomes/human/hg38-blacklist.v2.bed.gz\"),\n",
    "\to = f\"{PROJECT_DIR_r}genomes/human/excluded_regions.bed.gz\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aecad3af-bf80-4aeb-bd59-dd6bc1dead76",
   "metadata": {},
   "source": [
    "# Generate annotation for QC (gene body ratio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8816b3b9-2843-4a38-8219-e8012c59cf3c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T22:39:47.315044Z",
     "iopub.status.busy": "2024-10-11T22:39:47.313959Z",
     "iopub.status.idle": "2024-10-11T22:39:47.337529Z",
     "shell.execute_reply": "2024-10-11T22:39:47.336353Z",
     "shell.execute_reply.started": "2024-10-11T22:39:47.314982Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "RMSTemplate_generate_QC_gbratio_annotation_20240901 = load_func(pipeline, \"RMSTemplate_generate_QC_gbratio_annotation_20240901\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "cfe530f6-fc3d-4afd-a63f-7f3ae8772554",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-10-11T22:39:47.340546Z",
     "iopub.status.busy": "2024-10-11T22:39:47.339468Z",
     "iopub.status.idle": "2024-10-11T22:45:06.515699Z",
     "shell.execute_reply": "2024-10-11T22:45:06.514609Z",
     "shell.execute_reply.started": "2024-10-11T22:39:47.340490Z"
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "RMSTemplate_generate_QC_gbratio_annotation_20240901(\n",
    "\trms,\n",
    "\tconda_env = conda_env,\n",
    "\tannotation_file = f\"{PROJECT_DIR_r}genomes/human/gencode.v37.annotation.gtf.gz\",\n",
    "\to = f\"{PROJECT_DIR_d}QC/Annotations/gencode.v37.annotation_proteincoding_nonoverlapping_long_transcripts.gtf.gz\",\n",
    "\tdiscarded_chroms = [\"chrY\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "086ebed7-7796-4e28-a711-1dd651e452d9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
